### PR TITLE
e2e: Fixing compactor test flakiness.

### PR DESCRIPTION
**Pull request by bwplotka**
Also:

* Reduced number of services for e2e for latency
* Fixed halting
* Improved logging.
* Improved test cases (e.g added test for compaction and halting)

### CODE REVIEW

**Comment by bwplotka** [comment]
WIP, still iterating locally

**Comment by bwplotka** [comment]
Finally should be all working!

Added some TODOs to follow up, but not critical now, let's unblock CI: https://github.com/thanos-io/thanos/issues/2324

cc @kakkoyun @brancz

File: pkg/testutil/e2eutil/prometheus.go [hunk 1]
```diff
@@ -497,6 +469,10 @@ func createBlock(
 		return id, errors.Wrap(err, "write block")
 	}
 
+	if id.Compare(ulid.ULID{}) == 0 {
+		return id, errors.Errorf("nothing to write, asked for %v samples", numSamples)
```
Comment (yeya24) [inline pkg/testutil/e2eutil/prometheus.go]: We can use `%d` here.

File: test/e2e/compact_test.go [hunk 2]
```diff
@@ -30,530 +27,438 @@ import (
...
+			{Value: 240, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "2", "case": "partial-overlap-dedup-ready", "replica": "1"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "5", "case": "partial-overlap-dedup-ready", "replica": "1"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "2", "case": "partial-overlap-dedup-ready", "replica": "2"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "3", "case": "partial-overlap-dedup-ready", "replica": "2"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "2", "case": "partial-overlap-dedup-ready", "replica": "3"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "4", "case": "partial-overlap-dedup-ready", "replica": "3"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "1", "case": "partial-multi-replica-overlap-dedup-ready", "replica": "1", "rule_replica": "1"}},
+			{Value: 320, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "2", "case": "partial-multi-replica-overlap-dedup-ready", "replica": "1", "rule_replica": "1"}},
```
Comment (bwplotka) [inline test/e2e/compact_test.go]: This shows that we don't handle overlaps very well on store GW/Querier. it works, but it doesn't to dedup. Wonder if we can optimize?

File: test/e2e/compact_test.go [hunk 3]
```diff
@@ -30,530 +27,438 @@ import (
...
+		testutil.Ok(t, c.WaitSumMetrics(e2e.Equals(0), "thanos_compact_group_vertical_compactions_total"))
+		testutil.Ok(t, c.WaitSumMetrics(e2e.Equals(1), "thanos_compact_group_compactions_failures_total"))
+		testutil.Ok(t, c.WaitSumMetrics(e2e.Equals(2), "thanos_compact_group_compaction_runs_started_total"))
+		testutil.Ok(t, c.WaitSumMetrics(e2e.Equals(1), "thanos_compact_group_compaction_runs_completed_total"))
+
+		testutil.Ok(t, c.WaitSumMetrics(e2e.Equals(0), "thanos_compact_downsample_total"))
+		testutil.Ok(t, c.WaitSumMetrics(e2e.Equals(0), "thanos_compact_downsample_failures_total"))
+		testutil.Ok(t, s.Stop(c))
+	})
+	t.Run("native vertical deduplication shoukd kick in", func(t *testing.T) {
```
Comment (yeya24) [inline test/e2e/compact_test.go]: Small typo here `kick`

File: test/e2e/compact_test.go [hunk 4]
```diff
@@ -30,530 +27,438 @@ import (
...
+		testutil.Ok(t, str.WaitSumMetrics(e2e.Equals(0), "thanos_blocks_meta_sync_failures_total"))
+
+		testutil.Ok(t, c.WaitSumMetrics(e2e.Equals(0), "thanos_compactor_halted"))
+		// Make sure compactor does not modify anything else over time.
+		testutil.Ok(t, s.Stop(c))
+
+		ctx, cancel = context.WithTimeout(context.Background(), 3*time.Minute)
+		defer cancel()
+
+		// Check if query detects new blocks.
```
Comment (bwplotka) [inline test/e2e/compact_test.go]: duplicate comment, let's fix in another PR

File: test/e2e/compact_test.go [hunk 5]
```diff
@@ -30,530 +27,438 @@ import (
...
+		// Check if query detects current series, even if overlapped.
+		queryAndAssert(t, ctx, q.HTTPEndpoint(),
+			`count_over_time({a="1"}[12h])`,
```
Comment (bwplotka) [inline test/e2e/compact_test.go]: Probably we need 16.

```suggestion
			`count_over_time({a="1"}[16h])`,
```

File: pkg/testutil/e2eutil/prometheus.go [hunk 6]
```diff
@@ -345,91 +374,34 @@ func CreateBlockWithBlockDelay(
...
-	id, err = ulid.New(uint64(time.Unix(int64(blockID.Time()), 0).Add(-blockDelay*1000).Unix()), nil)
+	id, err := ulid.New(uint64(timestamp.FromTime(timestamp.Time(int64(blockID.Time())).Add(-blockDelay))), bytes.NewReader(blockID.Entropy()))
 	if err != nil {
-		return id, errors.Wrap(err, "create block id")
+		return ulid.ULID{}, errors.Wrap(err, "create block id")
 	}
 
 	if blockID.Compare(id) == 0 {
-		return
+		return ulid.ULID{}, errors.Wrap(err, "empty id, what the heck?")
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/testutil/e2eutil/prometheus.go]: oh, good point

**Pull request merged**